### PR TITLE
[browser][MT] Fix Promise cancelation

### PR DESCRIPTION
**Pull request by pavelsavara**
- fixed `compareExchangeI32` 🤦‍♂️
- unregister from `FinalizationRegistry` in `upgrade_managed_proxy_to_strong_ref`
- more testing if the runtime is still running
- testing if the promise holder is already disposed
- added `exit` to `RuntimeAPI`
- improve `BlazorHosted` WBT exit
- make `assert_runtime_running` always throw, in order to really protect the code after it.
   - make it not configurable by dropping `withAssertAfterExit`

Contributes to https://github.com/dotnet/runtime/issues/98721

### CODE REVIEW

File: src/mono/wasm/testassets/BlazorHostedApp/BlazorHosted.Client/Pages/Chat.razor [hunk 1]
```diff
@@ -86,7 +86,7 @@
     {
         await DisposeHubConnection();
         // exit the client
-        await JSRuntime.InvokeVoidAsync("eval", "import('./dotnet.js').then(module => { module.dotnet; module.exit(0); });");
+        Environment.Exit(0);
```
Comment (ilonatommy) [inline src/mono/wasm/testassets/BlazorHostedApp/BlazorHosted.Client/Pages/Chat.razor:89]: This does not produce a console message that we're waiting for to finish the task. 
https://github.com/dotnet/runtime/blob/ba13ef001d1f26c4de4741a74e70bd591cc901f2/src/mono/wasm/Wasm.Build.Tests/TestAppScenarios/AppTestBase.cs#L107
The test will never finish. Previous approach with `await JSRuntime.InvokeVoidAsync("eval", "getDotnetRuntime(0).exit(0);");` worked better.

**Review by ilonatommy — APPROVED** [review]

File: src/mono/wasm/testassets/BlazorHostedApp/BlazorHosted.Client/Pages/Chat.razor (hunk 1 above)
Comment (ilonatommy) [reply src/mono/wasm/testassets/BlazorHostedApp/BlazorHosted.Client/Pages/Chat.razor:89]: Or we can artificially print the message before it but it's a workaround.

Comment (maraf) [reply src/mono/wasm/testassets/BlazorHostedApp/BlazorHosted.Client/Pages/Chat.razor:89]: We can add `dotnet.withExitCodeLogging()` and do manual blazor startup

**Comment by maraf** [comment]
Please, filter out the noise in SignalR test log
appsettings.json in server project
```
{
  "Logging": {
    "LogLevel": {
      "Default": "Information",
      "Microsoft.AspNetCore": "Warning"
    }
  }
}
```

**Comment by pavelsavara** [comment]
We still have the issue here
[Log](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-99397-merge-2cead071929f4467b0/Workloads-NoWebcil-Wasm.Build.Tests.TestAppScenarios.SignalRClientTests/1/console.f13a36b8.log?helixlogtype=result)
```
[log] TestOutput -> Message = [LongPolling ping-pong]. ReceiveMessage from server on CurrentManagedThreadId=7
TestOutput -> Message = [LongPolling ping-pong]. ReceiveMessage from server on CurrentManagedThreadId=7
[] info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
[]       Request starting HTTP/1.1 POST http://localhost:5000/chathub?id=si6F6iWbbkYT2Dxjd4lpNQ - - 11
[] info: Microsoft.AspNetCore.Routing.EndpointMiddleware[0]
[]       Executing endpoint '/chathub'
[] info: Microsoft.AspNetCore.Routing.EndpointMiddleware[1]
[]       Executed endpoint '/chathub'
[] info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
[]       Request finished HTTP/1.1 POST http://localhost:5000/chathub?id=si6F6iWbbkYT2Dxjd4lpNQ - 200 0 text/plain 0.4975ms
[] info: Microsoft.AspNetCore.Routing.EndpointMiddleware[1]
[]       Executed endpoint '/chathub'
[] info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
[]       Request finished HTTP/1.1 GET http://localhost:5000/chathub?id=si6F6iWbbkYT2Dxjd4lpNQ - 200 11 application/octet-stream 145.1232ms
[error] [0x000dce2c--UI-] MONO_WASM: Assert failed: Expected Promise for GCHandle 374 Error: Assert failed: Expected Promise for GCHandle 374
    at gt (http://localhost:5000/_framework/dotnet.runtime.9.0.0-ci.pwdysoh9c6.js:3:20397)
    at lu (http://localhost:5000/_framework/dotnet.runtime.9.0.0-ci.pwdysoh9c6.js:3:203997)
    at http://localhost:5000/_framework/dotnet.native.wasm:wasm-function[8601]:0x212457
    at http://localhost:5000/_framework/dotnet.native.wasm:wasm-function[8645]:0x214723
    at http://localhost:5000/_framework/dotnet.native.wasm:wasm-function[8651]:0x21490f
    at http://localhost:5000/_framework/dotnet.native.wasm:wasm-function[8645]:0x214723
    at http://localhost:5000/_framework/dotnet.native.wasm:wasm-function[8772]:0x21b7d3
    at http://localhost:5000/_framework/dotnet.native.9.0.0-ci.7mfi08yqz3.js:8:103164
    at callUserCallback (http://localhost:5000/_framework/dotnet.native.9.0.0-ci.7mfi08yqz3.js:8:102608)
    at checkMailbox (http://localhost:5000/_framework/dotnet.native.9.0.0-ci.7mfi08yqz3.js:8:103143)
```

File: src/mono/wasm/testassets/BlazorHostedApp/BlazorHosted.Client/Pages/Chat.razor (hunk 1 above)
Comment (pavelsavara) [reply src/mono/wasm/testassets/BlazorHostedApp/BlazorHosted.Client/Pages/Chat.razor:89]: I added `withExitCodeLogging`

**Pull request merged**